REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [6]:
len(bus['business_id'].unique())

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [7]:
bus[bus['city']=='San Francisco']['name'].value_counts().head(5)

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [8]:
bus['postal_code'].values

array(['94104', '94109', '94104', ..., '94116', '94112', '94110'],
      dtype=object)

In [9]:
type(bus['postal_code'].values[0])

str

#### Question 3b

What are the unique values of postal_code?

In [10]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [11]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [12]:
vb = bus[bus['postal_code'].isin(validZip)]

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [13]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [14]:
null_lon = vb[pd.isnull(vb['longitude'])]['postal_code'].value_counts()
not_null_lon = vb[pd.notnull(vb['longitude'])]['postal_code'].value_counts()
df = pd.merge(null_lon,not_null_lon,left_index=True, right_index=True).reset_index()
df.columns = ['postal_code','null_lon','not_null_lon']
df

postal_code  null_lon  not_null_lon
0        94103       252           306
1        94110       251           334
2        94107       247           204
3        94102       196           262
4        94109       146           237
5        94133       135           282
6        94122       123           147
7        94111       115           167
8        94114       108           122
9        94105       100           126
10       94118       100           143
11       94124        88            76
12       94108        82           142
13       94115        76           162
14       94117        67           130
15       94132        65            69
16       94104        59            74
17       94112        59           131
18       94121        55           106
19       94123        53           116
20       94116        35            58
21       94134        29            45
22       94127        28            43
23       94131        14            34

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [15]:
df['null_ratio'] = df['null_lon']/(df['null_lon']+df['not_null_lon'])
df.sort_values(['null_ratio'],ascending=False).head(5)

postal_code  null_lon  not_null_lon  null_ratio
2        94107       247           204    0.547672
11       94124        88            76    0.536585
15       94132        65            69    0.485075
8        94114       108           122    0.469565
6        94122       123           147    0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [30]:
print(len(ins['business_id'].unique()))
print(len(bus['business_id'].unique()))
print(len(ins))
print(len(bus))

5730
6315
15430
6315


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [17]:
type_list = ins['type'].value_counts()[ins['type'].value_counts()>=10].index
ins2 = ins[ins['type'].isin(type_list)]
ins2

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

In [18]:
ins2['type'].value_counts()

routine    15429
Name: type, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [31]:
ins['year'] = pd.to_datetime(ins.date,format='%Y%m%d').dt.year
ins

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15430 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [20]:
ins['year'].unique()

array([2016, 2014, 2015, 2013], dtype=int64)

In [21]:
ins.groupby(['year'])['business_id'].count()

year
2013      38
2014    5629
2015    3923
2016    5840
Name: business_id, dtype: int64

In [22]:
year_list = ins.groupby(['year'])['business_id'].count()[ins.groupby(['year'])['business_id'].count()>50].index
ins3 = ins[ins['year'].isin(year_list)]
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15392 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [33]:
ins3[ins3['year']==2016].groupby(['business_id']).count()['score']

business_id
10       1
19       1
24       2
31       1
45       2
        ..
89515    1
89701    1
89790    1
89803    1
89900    1
Name: score, Length: 4543, dtype: int64

In [35]:
ins2016=ins3[ins3['year']==2016]
ins2016

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5840 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [37]:
bus_ins = pd.merge(bus,ins2016,on='business_id',how='left')
bus_ins.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score        date     type    year  
0 -122.403816  +14154217044   82.0  20160503.0  routine  2016.0  
1 -122.421547  +14157763262   94.0  20160513.0  routine  2016.0  
2 -122.403135  +14156779494   98.0  20161005.0  routine  2016.0  
3 -122.403135  +14156779494   96.0  20160311.0  routine  2016.0  
4 -122.419004           NaN   84.0  20160816.0  routine  2016.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [38]:
bus_ins.sort_values(['score'],ascending=True).head(20)

business_id                         name              address  \
4892        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    
7325        87652                    POKI TIME     2101 LOMBARD ST    
6281        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    
4873        71232                   UNCLE CAFE       65 WAVERLY PL    
3008        34585             HING WANG BAKERY        339 JUDAH ST    
5627        78328                   GOLDEN WOK   295 B ORIZABA AVE    
5383        76615                   TOP'S CAFE  2460 SAN BRUNO AVE    
6067        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    
836          2716                 RED A BAKERY      634 CLEMENT ST    
4209        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    
549          1843          MA'S DIM SUM & CAFE      1315 POWELL ST    
4903        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    
6645        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    
624          2044          JOANIE'S DINER INC.   1329 COLUMBUS AVE    
6938        85587         PAK NAM THAI CUISINE       655 LARKIN ST    
3563        61161                  JADE GARDEN      3608 BALBOA ST    
4234        67564                        ITHAI         720 POST ST    
6151        81106          VALLARTA'S TACO BAR     2022 MISSION ST    
4850        70996             HONG KONG LOUNGE     5322 GEARY BLVD    
5013        72176                 WOLFES LUNCH        1220 16TH ST    

               city state postal_code   latitude   longitude  phone_number  \
4892  San Francisco    CA       94121  37.780144 -122.482040  +14156685105   
7325  San Francisco    CA       94123        NaN         NaN  +14156366868   
6281  San Francisco    CA       94112        NaN         NaN  +14157348687   
4873  San Francisco    CA       94108  37.794034 -122.406794  +14155701380   
3008  San Francisco    CA       94122  37.762174 -122.465809  +14156813928   
5627  San Francisco    CA       94132        NaN         NaN  +14155868000   
5383  San Francisco    CA       94134        NaN         NaN  +14157156868   
6067  San Francisco    CA       94109        NaN         NaN  +14158856402   
836   San Francisco    CA       94118  37.782888 -122.466101  +14155663888   
4209  San Francisco    CA       94122  37.763891 -122.468805  +14152420828   
549   San Francisco    CA       94133  37.796683 -122.410134  +14157883532   
4903  San Francisco    CA       94108        NaN         NaN  +14159567888   
6645  San Francisco    CA       94123        NaN         NaN  +14156665567   
624   San Francisco    CA       94133  37.806546 -122.419277  +14158232670   
6938  San Francisco    CA       94109        NaN         NaN  +14156097089   
3563  San Francisco    CA       94121  37.775728 -122.497612  +14152218168   
4234  San Francisco    CA       94109  37.787691 -122.413717  +14156581010   
6151  San Francisco    CA       94110        NaN         NaN  +14152528810   
4850  San Francisco    CA       94121  37.780538 -122.476632           NaN   
5013  San Francisco    CA       94107        NaN         NaN           NaN   

      score        date     type    year  
4892   52.0  20161031.0  routine  2016.0  
7325   55.0  20160721.0  routine  2016.0  
6281   55.0  20160901.0  routine  2016.0  
4873   55.0  20160308.0  routine  2016.0  
3008   56.0  20160201.0  routine  2016.0  
5627   56.0  20160824.0  routine  2016.0  
5383   57.0  20160504.0  routine  2016.0  
6067   57.0  20160129.0  routine  2016.0  
836    57.0  20160323.0  routine  2016.0  
4209   57.0  20160825.0  routine  2016.0  
549    57.0  20160216.0  routine  2016.0  
4903   57.0  20161121.0  routine  2016.0  
6645   58.0  20160714.0  routine  2016.0  
624    59.0  20160204.0  routine  2016.0  
6938   59.0  20160927.0  routine  2016.0  
3563   59.0  20160107.0  routine  2016.0  
4234   59.0  20161108.0  routine  2016.0  
6151   60.0  20160811.0  routine  2016.0  
4850   60.0  20160121.0  routine  2016.0  
5013   60.0  20

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!